File: lib/parser.rb
require 'pry'

class Parser
  class UnexpectedToken < StandardError; end
  class UnexpectedEOFWhileReading < StandardError; end
  class << self
    def execute(code:)
      read_from(tokenize(code))
    end

    private

    def read_from(tokens)
      raise UnexpectedEOFWhileReading if tokens.length == 0
      token = tokens.shift
      case token
      when '('
        list = []
        while tokens[0] != ')'
          list << read_from(tokens)
        end
        tokens.shift
        list
      when ')'
        tokens.shift
        read_from(tokens)
      else
        atom(token)
        end
    end

    def tokenize(s)
      s.gsub('(', ' ( ').gsub(')', ' ) ').split(' ')
    end

    def atom(token)
      if token.match(%r{^\d+$})
        token.to_i
      elsif token.match(%r{\".*\"})
        token.gsub('"', '')
      else
        token.to_sym
      end
    end
  end
end

File: erl/parser.erl
-module(parser).
-export([execute/1, read_from/1, while/1, to_atom/1]).

execute(Token) ->
		Tokens = compact(Token),
		Tokens.
		%% read_from(Tokens).

tokenize(Token)->
		S = re:replace(Token, "\\(", " ( ", [global, {return, list}]),
		R = re:replace(S, "\\)", " ) ", [global, {return, list}]),
		re:split(R, " ", [{return, list}]).

compact(Token) ->
		T = tokenize(Token),
		[L || L <- T, length(L) >= 1].

to_atom(Value) ->
		{_, CondFloat} = re:compile("[0-9]+\.[0-9]+"),
		RegFloat = re:run(Value, CondFloat),
		case RegFloat of
				{match, _} ->
						hd(string:to_float(Value))
		end.

type_translate(Token) ->
		[to_atom(Value) || Value <- Token].

read_from(X) ->
		type_translate(X).

while(X) ->
		X.

%% read_from(Tokens) ->
%%  		[Car|Cdr] = Tokens,
%%  		read_from(Car, Cdr).
%%  
%% read_from("(", Cdr) ->
%%  		while(Cdr);
%% read_from(")", Cdr) ->
%%  		read_from(Cdr);
%% read_from(Car, _) ->
%%  		[Car].
%%  
%% while(List) ->
%%  		[H|T] = List,
%%  		L = [],
%%  		while(H,T,L).
%%  
%% while(H, _, L) when H == ")" ->
%%  		L;
%% while(H, T, L) ->
%%  		[Car|Cdr] = T,
%%  		while(Car, Cdr, lists:append(L, [H])).
